src/treelearner/serial_tree_learner.h

```diff
@@ -215,6 +286,17 @@ class SerialTreeLearner: public TreeLearner {
   const Json* forced_split_json_;
   std::unique_ptr<TrainingShareStates> share_state_;
   std::unique_ptr<CostEfficientGradientBoosting> cegb_;
+  /*! \brief whether numerical features contain any nan values, used for linear model */
+  std::vector<int8_t> contains_nan_;
+  /*! whether any numerical feature contains a nan value, used for linear model */
+  bool any_nan_;
+  /*! \brief map dataset to leaves, used for linear model */
+  std::vector<int> leaf_map_;
```

Comment:
is this different with DataPartition?

Reply:
Yes, I think so. `DataPartition` stores the data indices sorted by leaf number. `leaf_map_` stores the leaf number of each data index, with the data indices in original order.